[#IP-86] tslint to eslint migration

michaeldisaro opened the pull request:
#### List of Changes
- Deleted `tslint.json` file
- Removed tslint packages from `package.json`
- Created `.eslintignore` ignore file
- Created `.eslintrc.js` configuration
- Added some lint exclusions to `.eslintrc.js` to avoid a deep refactoring of working code
- Changed tslint's "rule disabling" comments with eslint closest ones.

#### Motivation and Context
We want to migrate to eslint because tslint is deprecated.

#### How Has This Been Tested?
It has been tested by performing:
- yarn install --frozen-lockfile
- yarn build
- yarn lint
- yarn test

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

BurnedMarshal commented on .eslintignore:
```diff
@@ -0,0 +1,9 @@
+# Never lint node_modules
```
This file could be removed. We have defined exclusion in `ignorePatterns` and these settings are redundant.

michaeldisaro replied:
Ok, fine.

BurnedMarshal commented on package.json:
```diff
@@ -14,9 +14,11 @@
     "start": "node dist/index.js",
     "test": "dotenv -e env.example -- jest",
-    "lint": "tslint --project ."
+    "lint": "eslint . -c .eslintrc.js --ext .ts,.tsx"
```
you could add a fix linting script
```suggestion
    "lint": "eslint . -c .eslintrc.js --ext .ts,.tsx",
    "lint:fix": "eslint . -c .eslintrc.js --ext .ts,.tsx --fix"
```

michaeldisaro replied:
I talked with @balanza about that, we think that "yarn lint --fix" is less verbose than the "yarn lint-autofix" that I've seen somewhere else. Even if "yarn lint:fix" is effectively less verbose than the others, I think that it's more natural to add the "--fix" when launching yarn lint because the command itself suggests this. What do you think?

BurnedMarshal replied:
Nice idea. We can keep the standard `yarn lint --fix`. The custom script inside the package json is useful only as a suggestion for contributors.

michaeldisaro replied:
Contributors can see the lint message that says to ass --fix too. Let's keep it simple.

BurnedMarshal commented on .eslintrc.js:
```diff
@@ -0,0 +1,36 @@
+        "@pagopa/eslint-config/strong",
+    ],
+    "rules": {
+        "functional/prefer-readonly-type": "off",
+        "prefer-arrow/prefer-arrow-functions": "off",
+        "@typescript-eslint/no-use-before-define": "off",
+        "import/order": "off",
+        "@typescript-eslint/no-floating-promises": "off",
+        "prettier/prettier": "off",
+        "sort-keys": "off",
+        "@typescript-eslint/semi": "off",
+        "@typescript-eslint/explicit-member-accessibility": "off",
+        "jsdoc/newline-after-description": "off",
+        "arrow-body-style": "off"
```
We can enable almost all this rules using auto fix eslint feature.
The following rules can remain disabled if require a deep refactory.
```
"prefer-arrow/prefer-arrow-functions": "off",
"sort-keys": "off",
"@typescript-eslint/no-use-before-define": "off",
```

balanza replied:
Or, we can disable them locally so that at least new implementations will cope with standard rules

balanza replied:
Example here: https://github.com/pagopa/io-functions-express/pull/16#pullrequestreview-628731997

BurnedMarshal commented on src/index.ts:
```diff
@@ -54,5 +54,5 @@ tryCatch(() => emailService.verifyTransport(), toError)
       })
   )
-  // tslint:disable-next-line:no-floating-promises
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
```
This line should be moved at the top of `tryCatch` chain to work properly.

michaeldisaro replied:
Done.

michaeldisaro commented:
@BurnedMarshal @balanza I fixed broken tests, removed global ignores and made some of your requested fixes. Let me know if we need anything else to merge. Tests passed and build works.

BurnedMarshal commented on src/utils/environment.ts:
```diff
@@ -27,8 +28,8 @@ export function getRequiredEnvVar(varName: string): string | never {
+export const readFile = (path: string, type: string): string => {
   log.info("Reading %s file from %s", type, path);
-  // eslint-disable-next-line security/detect-non-literal-fs-filename
+  // eslint-disable-next-line
```
This line seems not required

balanza replied:
@michaeldisaro in general, avoid disabling eslint totally - always be specific on which rule you're going to disable

michaeldisaro replied:
ok, I fixed that behavior on my tool.

michaeldisaro replied:
I removed the line, it was effectively useless.

BurnedMarshal reviewed (approved):
LGTM apart from a minor change. Thanks @michaeldisaro

balanza reviewed (approved):
LGTM apart minor, non blocking comments
